src/output/datafrog_opt.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::time::Instant;
+
+use crate::facts::{AllFacts, Loan, Point, Region};
+use crate::output::Output;
+
+use datafrog::{Iteration, Relation};
+
+pub(super) fn compute(dump_enabled: bool, mut all_facts: AllFacts) -> Output {
+    // Declare that each universal region is live at every point.
+    let all_points: BTreeSet<Point> = all_facts
+        .cfg_edge
+        .iter()
+        .map(|&(p, _)| p)
+        .chain(all_facts.cfg_edge.iter().map(|&(_, q)| q))
+        .collect();
+
+    for &r in &all_facts.universal_region {
+        for &p in &all_points {
+            all_facts.region_live_at.push((r, p));
+        }
+    }
+
+    let timer = Instant::now();
+
+    let mut result = Output::new(dump_enabled);
+
+    let borrow_live_at = {
+        // Create a new iteration context, ...
+        let mut iteration = Iteration::new();
+
+        // .. some variables and different indices for `subset`.
+        let subset = iteration.variable::<(Region, Region, Point)>("subset");
+        let subset_1 = iteration.variable_indistinct("subset_1");
+        let subset_2 = iteration.variable_indistinct("subset_2");
+        let subset_r1p = iteration.variable_indistinct("subset_r1p");
+        let subset_r2p = iteration.variable_indistinct("subset_r2p");
+        let subset_p = iteration.variable_indistinct("subset_p");
+
+        // temporaries as we perform a multi-way join, and more indices
+        let requires = iteration.variable::<(Region, Loan, Point)>("requires");
+        let requires_1 = iteration.variable("requires_1");
+        let requires_2 = iteration.variable("requires_2");
+        let requires_bp = iteration.variable("requires_bp");
+        let requires_rp = iteration.variable("requires_rp");
```

Comment:
I apologize for not noticing this before, but each of the `requires_` variables can be `variable_indistinct()`. We just need `requires` itself (without a `_`) to be a `variable()` without `_indistinct()`. This has a solid perf benefit for me, taking the time from 9.43s to 7.99s.

Reply:
Solid benefits for me as well, 10-15% 👍